### ISSUE TITLE
How would one use this?

### ISSUE THREAD

**Issue by Xynonners** [AaronSpieler/elmneuron#1]
Hi there,

I read the paper on arxiv and it seems pretty interesting. My question is, how would one use this to solve traditional ML problems like the ones demonstrated in LRA?

**Comment by AaronSpieler** [comment]
Hey, 

thanks for your interest! 

Example code/notebook of how to solve machine learning tasks (e.g. SHD) you can find [here](https://github.com/AaronSpieler/elmneuron/blob/main/notebooks/train_elm_on_shd.ipynb). If you question is specifically about LRA we detail the training methodology in the appendix of the paper.

If you are generally interested in how to deal with 1D sequences/time-series data: I would recommend to have a one-hot-encoding, embedding, or a projection layer first, and then apply the ELM neuron on top. In the paper we did something like coarser one-hot-encoding, but in retrospect it would have worked probably much better to use an embedding layer. 

The key thing to note is that the ELM neuron has the biggest computational benefit on sparse large dimensional input data (as far as I know), i.e. where `d_inp >> dim(num_memory)`, assuming `num_memory` is sufficient to solve the task well. So datasets much closer to NeuronIO (`d_inp=1278`), SHD (`d_inp=700`), SHD-Adding (`d_inp=700`) or Audio data (typicall large `d_inp`) might be a much better fit. 

In that sense something like a 1D sequence is quite adversarial to the ELM neuron architecture, but can be made to work, and we choose LRA to highlight the long range processing capabilities of the architecture.

For long-range processing capabilities, just adjust the maximal memory timescale (`memory_tau_max`) hyper-parameter to your sequence length, and for temporally very sparse data, additionally adjust the synapse timescale (`tau_b_value`) to the expected time interval between two inputs (rarely neeced though).

Last note on input type: we mostly evaluated on integer valued input, but float values can also works fine, however, if you are increasing the branch size (`num_synapse_per_branch`), then you should ideally not mix positive and negative input (they can cancel out in the linear sum), but stick to either positive or negative input.

Good luck!
Aaron
